core/src/main/kotlin/xyz/block/artifactswap/core/module_selector/ArtifactSwapModuleSelector.kt

```diff
@@ -0,0 +1,267 @@
+package xyz.block.artifactswap.core.module_selector
+
+import com.fueledbycaffeine.spotlight.buildscript.GradlePath
+import kotlin.collections.map
+import kotlin.coroutines.CoroutineContext
+import kotlin.time.Duration
+import kotlin.time.measureTimedValue
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+import xyz.block.artifactswap.core.download.services.ArtifactSyncBomLoader
+import xyz.block.artifactswap.core.eventstream.Eventstream
+
+private val LOGGER = LoggerFactory.getLogger(RealArtifactSwapModuleSelector::class.java)
+
+/** Result of project selection with metrics about why projects were included. */
+data class ModuleSelectionResult(
+  val selectedProjects: Set<GradlePath>,
+  val metrics: SelectionMetrics,
+  val bomVersion: String,
+  val event: ModuleSelectionEvent,
+)
+
+/** Metrics about project selection decisions. */
+data class SelectionMetrics(
+  val totalCandidates: Int,
+  val totalSelected: Int,
+  val selectedDueToExplicitRequest: Int,
+  val selectedDueToAlwaysKeep: Int,
+  val selectedDueToLocalChanges: Int,
+  val selectedDueToMissingArtifact: Int,
+  val excludedDueToArtifactAvailable: Int,
+)
+
+/** Internal result data for module selection computation. */
+internal data class SelectionComputationResult(
+  val selectedProjects: Set<GradlePath>,
+  val metrics: SelectionMetrics,
+  val bomVersionToUse: String,
+  val bomDuration: Duration,
+  val gitDuration: Duration,
+  val locallyChangedFilesCount: Int,
+  val artifactsDuration: Duration,
+  val swappableArtifactsCount: Int,
+  val selectionDuration: Duration,
+)
+
+interface ArtifactSwapModuleSelector {
+  fun selectProjects(
+    candidates: Set<GradlePath>,
+    requestedProjects: Set<GradlePath>,
+  ): ModuleSelectionResult
+}
+
+class RealArtifactSwapModuleSelector(
```

Comment:
This code is largely the same as `ArtifactSyncImportFileHelper` but i broke up the implementation for readability